Refactor `Button` component to TypeScript

kienstra opened the pull request:
(Ready for review)

## What?
Convert the `Button` component to TypeScript

## Why?
As part of an [effort](https://github.com/WordPress/gutenberg/issues/35744) to convert `@wordpress/components` to TypeScript

## How?
Mainly by adding types to the `Button` component

## Testing Instructions
1. `npm run storybook:dev`
2. Expected: The stories work like before

## Screenshots

![Screen Shot 2023-01-15 at 5 27 38 PM](https://user-images.githubusercontent.com/4063887/212573342-ecdc7908-cfe2-4f42-8c81-497c17b16880.png)

kienstra committed:
Rename button/index.js to .jsx

kienstra committed:
Possible approach to the prop types

kienstra committed:
Simplify the tag container

kienstra committed:
Allow passing childrent to TagElement

kienstra committed:
Fix the type of onKeyDown

kienstra committed:
Rename test/index.js to .tsx, fix some errors

kienstra committed:
Fix an eslint error from children being defined in the outer scope

kienstra committed:
Rename stories/index.js to .tsx, add type annotations

kienstra committed:
Merge branch 'trunk' into update/button-ts

kienstra committed:
Alphabetize types.ts types

kienstra committed:
Refactor TagElement to element

kienstra committed:
Alphabetize an import

kienstra committed:
Revert changes to element, this will fail type checking

kienstra committed:
Merge branch 'trunk' into update/button-ts

kienstra committed:
Use the [] array type syntax

kienstra committed:
Improve the typing of hasChildren

kienstra committed:
Fix the typing of chilren.length

kienstra committed:
Revert "Revert changes to element, this will fail type checking"

This reverts commit 0b616c785677c407879b6b0f7708409b6ca3dbaf.

kienstra committed:
Pass the generic of WordPressComponentProps a second argument

kienstra committed:
Don't destructure out children, simply let them be in props

kienstra commented on packages/components/src/border-control/border-control-dropdown/component.tsx:
```diff
@@ -166,7 +166,7 @@ const BorderControlDropdown = (
 			variant="tertiary"
 			aria-label={ toggleAriaLabel }
-			position={ dropdownPosition }
+			tooltipPosition={ dropdownPosition }
```
`<Button>` has no `position` prop, so maybe this should be `tooltipPosition`. 

This also has a `showTooltip` prop below.

kienstra commented on packages/components/src/dropdown/stories/index.tsx:
```diff
@@ -44,7 +44,7 @@ const Template: ComponentStory< typeof Dropdown > = ( args ) => {
 Default.args = {
 	renderToggle: ( { isOpen, onToggle } ) => (
-		<Button onClick={ onToggle } aria-expanded={ isOpen } isPrimary>
+		<Button onClick={ onToggle } aria-expanded={ isOpen } variant="primary">
```
The `isPrimary` prop is deprecated.

kienstra commented on packages/components/src/form-token-field/token.tsx:
```diff
@@ -73,7 +73,7 @@ export default function Token( {
 				className="components-form-token-field__remove-token"
 				icon={ closeSmall }
-				onClick={ ! disabled && onClick }
+				onClick={ ! disabled ? onClick : undefined }
```
`onClick` doesn't accept `false`, but it does accept `undefined`

kienstra commented on packages/components/src/tab-panel/index.tsx:
```diff
@@ -25,7 +25,7 @@ const TabButton = ( {
 	<Button
 		role="tab"
-		tabIndex={ selected ? null : -1 }
+		tabIndex={ selected ? undefined : -1 }
```
Sorry, this expands the scope outside of `components/src/button/`.

But `tabIndex` doesn't accept `null`, only `number | undefined`

kienstra commented on packages/components/src/tab-panel/types.ts:
```diff
@@ -34,7 +35,7 @@ type Tab< IconProps = unknown > = {
 	children: ReactNode;
 	label?: string;
-	onClick: ( event: MouseEvent ) => void;
+	onClick: Parameters< typeof Button >[ 0 ][ 'onClick' ];
```
Corrects a TS error:

```
packages/components/src/tab-panel/index.tsx:26:3 - error TS2322: Type '{ children: ReactNode; label?: string | undefined; onClick: (event: MouseEvent) => void; showTooltip?: boolean | undefined; className?: string | undefined; icon?: IconType<...> | undefined; ... 5 more ...; __experimentalIsFocusable: true; }' is not assignable to type 'Omit<Pick<DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>, "key" | keyof ButtonHTMLAttributes<...>> | Pick<...>, "as" | keyof ButtonProps>'.
  Types of property 'onClick' are incompatible.
    Type '(event: MouseEvent) => void' is not assignable to type 'MouseEventHandler<HTMLButtonElement> | MouseEventHandler<HTMLAnchorElement> | undefined'.
      Type '(event: MouseEvent) => void' is not assignable to type 'MouseEventHandler<HTMLButtonElement>'.
        Types of parameters 'event' and 'event' are incompatible.
          Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is missing the following properties from type 'MouseEvent': offsetX, offsetY, x, y, and 14 more.
```

Uses the `Parameters` generic instead of `WordPressComponentProps< ButtonProps, TagName >[ 'onClick' ]`

kienstra committed:
Use React.MousEvent

kienstra commented on packages/components/src/tab-panel/types.ts:
```diff
@@ -1,7 +1,7 @@
  * External dependencies
  */
-import type { ReactNode } from 'react';
+import type { MouseEvent, ReactNode } from 'react';
```
Fixes a TS error in [tab-panel/index.tsx#L26](https://github.com/WordPress/gutenberg/blob/e702dd475b9c8eba500f9fd2eea58c03bc7f728d/packages/components/src/tab-panel/index.tsx#L26):

```
packages/components/src/tab-panel/index.tsx:26:3 - error TS2322: Type '{ children: ReactNode; label?: string | undefined; onClick: (event: MouseEvent) => void; showTooltip?: boolean | undefined; className?: string | undefined; icon?: IconType<...> | undefined; ... 5 more ...; __experimentalIsFocusable: true; }' is not assignable to type 'Omit<Pick<DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>, "key" | keyof ButtonHTMLAttributes<...>> | Pick<...>, "as" | keyof ButtonProps>'.
  Types of property 'onClick' are incompatible.
    Type '(event: MouseEvent) => void' is not assignable to type 'MouseEventHandler<HTMLButtonElement> | MouseEventHandler<HTMLAnchorElement> | undefined'.
      Type '(event: MouseEvent) => void' is not assignable to type 'MouseEventHandler<HTMLButtonElement>'.
        Types of parameters 'event' and 'event' are incompatible.
          Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is missing the following properties from type 'MouseEvent': offsetX, offsetY, x, y, and 14 more.

26  <Button
```

This uses `React.MouseEvent` instead of `MouseEvent` in [tab-panel/types.ts#L37](https://github.com/WordPress/gutenberg/blob/e702dd475b9c8eba500f9fd2eea58c03bc7f728d/packages/components/src/tab-panel/types.ts#L37).

kienstra commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -273,7 +274,7 @@ type DayProps = {
 	isInvalid: boolean;
 	onClick: () => void;
-	onKeyDown: ( event: KeyboardEvent ) => void;
+	onKeyDown: KeyboardEventHandler;
```
`onKeyDown` is passed to `Button`, and this caused type errors before:

```
packages/components/src/date-time/date/index.tsx:317:4 - error TS2322: Type '(event: KeyboardEvent) => void' is not assignable to type 'KeyboardEventHandler<HTMLAnchorElement> | KeyboardEventHandler<HTMLButtonElement> | undefined'.
  Type '(event: KeyboardEvent) => void' is not assignable to type 'KeyboardEventHandler<HTMLAnchorElement>'.
    Types of parameters 'event' and 'event' are incompatible.
      Type 'KeyboardEvent<HTMLAnchorElement>' is missing the following properties from type 'KeyboardEvent': isComposing, initKeyboardEvent, DOM_KEY_LOCATION_LEFT, DOM_KEY_LOCATION_NUMPAD, and 14 more.

317    onKeyDown={ onKeyDown }
       ~~~~~~~~~

  node_modules/@types/react/index.d.ts:1421:9
    1421         onKeyDown?: KeyboardEventHandler<T> | undefined;
                 ~~~~~~~~~
    The expected type comes from property 'onKeyDown' which is declared here on type 'IntrinsicAttributes & ButtonProps & Omit<Pick<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>, "key" | keyof AnchorHTMLAttributes<...>> | Pick<...>, "as" | keyof ButtonProps> & { ...; } & RefAttributes<...> & { ...; } & { ...; }'


Found 1 error.
```

kienstra commented on packages/components/src/border-box-control/border-box-control-linked-button/component.tsx:
```diff
@@ -27,7 +27,7 @@ const BorderBoxControlLinkedButton = (
 			<View className={ className }>
 				<Button
-					{ ...buttonProps }
+					{ ...( buttonProps as Parameters< typeof Button >[ 0 ] ) }
```
Sorry to expand the scope outside of `components/src/button/`

This corrects a TS error:

```
ackages/components/src/border-box-control/border-box-control-linked-button/component.tsx:29:6 - error TS2322: Type '{ isSmall: true; icon: Element; iconSize: number; "aria-label": string; ref: ForwardedRef<any>; onClick: () => void; onChange?: FormEventHandler<HTMLDivElement> | undefined; ... 250 more ...; as?: keyof IntrinsicElements | undefined; }' is not assignable to type 'Omit<Pick<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>, "key" | keyof AnchorHTMLAttributes<...>> | Pick<...>, "as" | keyof ButtonProps>'.
  Types of property 'onChange' are incompatible.
    Type 'FormEventHandler<HTMLDivElement> | undefined' is not assignable to type 'FormEventHandler<HTMLAnchorElement> | FormEventHandler<HTMLButtonElement> | undefined'.
      Type 'FormEventHandler<HTMLDivElement>' is not assignable to type 'FormEventHandler<HTMLAnchorElement> | FormEventHandler<HTMLButtonElement> | undefined'.
        Type 'FormEventHandler<HTMLDivElement>' is not assignable to type 'FormEventHandler<HTMLAnchorElement>'.
          Type 'HTMLDivElement' is missing the following properties from type 'HTMLAnchorElement': charset, coords, download, hreflang, and 21 more.

```

mirka replied:
Not at all, you caught a legitimate typing error here 🙂

The error is coming from the fact that the props for `BorderBoxControlLinkedButton` (and also `useBorderBoxControlLinkedButton`) are typed as if the rest props would forward to a `div`, when in fact they forward to a `button`.

https://github.com/WordPress/gutenberg/blob/16f74d32944e1779be211f8d8b91ea2c06492ffc/packages/components/src/border-box-control/border-box-control-linked-button/component.tsx#L19

☝️ So we should fix these to `'button'` instead of casting `buttonProps` with `as`.

kienstra replied:
Great idea, committed in https://github.com/WordPress/gutenberg/pull/46997/commits/31cbdab15c013bae56a08cedbc132439a1eb6b8d

kienstra commented on packages/components/src/button/index.tsx:
```diff
@@ -93,10 +106,12 @@ export function Button( props, ref ) {
+			children?.[ 0 ] &&
+			children[ 0 ] !== null &&
+			// Tooltip should not considered as a child
+			children?.[ 0 ]?.props?.className !== 'components-tooltip' );
```
This is more complex than I'd like.

But simply checking for `Array.isArray( children )` before `children?.[ 0 ]` wasn't enough, and it caused unit tests to fail.

This is because `children` can be a `string`, which also has a `.length` property.

kienstra commented on packages/components/src/button/index.tsx:
```diff
@@ -171,9 +190,30 @@ export function Button( props, ref ) {
+			>
+				{ elementChildren }
+			</button>
+		);
```
`const element` is really verbose.

But [without this](https://github.com/WordPress/gutenberg/blob/0b616c785677c407879b6b0f7708409b6ca3dbaf/packages/components/src/button/index.tsx#L174), there are type errors:

```
const Tag: "a" | "button"
Type '{ children: (ReactNode | Element)[]; className: string; "aria-label": string | undefined; "aria-describedby": string | undefined; ref: ForwardedRef<any>; ... 356 more ...; wrap?: string | undefined; }' is not assignable to type 'AnchorHTMLAttributes<HTMLAnchorElement>'.
  Types of property 'onCopy' are incompatible.
    Type 'ClipboardEventHandler<HTMLAnchorElement> | ClipboardEventHandler<HTMLButtonElement> | ClipboardEventHandler<...> | undefined' is not assignable to type 
```

![Screen Shot 2023-01-15 at 4 25 28 PM](https://user-images.githubusercontent.com/4063887/212570500-da013218-bbde-42e2-aac3-85b3a463f38b.png)

kienstra commented on packages/components/src/button/test/index.tsx:
```diff
@@ -360,28 +350,33 @@ describe( 'Button', () => {
 	describe( 'deprecated props', () => {
 		it( 'should not break when the legacy isPrimary prop is passed', () => {
+			// @ts-expect-error
 			render( <Button isPrimary /> );
```
The `isPrimary` prop is deprecated, so there should be a TS error.

kienstra committed:
Update README.md with types from types.ts

kienstra committed:
Fix the type of shortcut

kienstra committed:
Remove unnecessary Type annotations

kienstra committed:
In README.md, alphabetize the props

kienstra committed:
Add a Required annotation to the variant prop

kienstra committed:
Add a JS DocBlock to Button

kienstra commented on packages/components/src/button/index.tsx:
```diff
@@ -113,7 +128,7 @@ export function Button( props, ref ) {
 	const trulyDisabled = disabled && ! isFocusable;
 	const Tag = href !== undefined && ! trulyDisabled ? 'a' : 'button';
-	const tagProps =
+	const tagProps: HTMLProps< TagName > =
```
This typing looks unnecessary, but it prevents a TS error below.

kienstra committed:
Add a CHANGELOG entry

kienstra commented on packages/components/src/button/test/index.tsx:
```diff
@@ -139,16 +139,6 @@ describe( 'Button', () => {
 		} );
 
-		it( 'should render an additional WordPress prop of value awesome', () => {
-			render( <Button WordPress="awesome" /> );
```
Now that `Button` is in TypeScript, it can't accept an arbitrary prop like `WordPress`

mirka replied:
I couldn't glean any additional hints from the [original PR](https://github.com/WordPress/gutenberg/pull/946/files#diff-49d4d17aab9174c55b8ff29058dbc2fa5852e913e5afe9104d24ce0e0a42c5d4R63) that introduced this test, but I'd assume its intent is to assert that rest props are passed down to the underlying element. We can change the test case to something valid like `<Button type="submit">`.

kienstra replied:
Good idea, committed in https://github.com/WordPress/gutenberg/pull/46997/commits/35f8efb841a4567f5204edeaf9556d5c6ad3df1e

mirka commented on packages/components/src/button/types.ts:
```diff
@@ -0,0 +1,125 @@
+	 * If provided with `showTooltip`, sets the position of the tooltip.
+	 * Please refer to the Tooltip component for more details regarding the defaults.
+	 */
+	tooltipPosition?: string;
```
Let's reference the types for the underlying component:

```suggestion
	tooltipPosition?: PopoverProps[ 'position' ];
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/46997/commits/916f9ac45ae0d40b844bc58884828e9077601d40

mirka commented on packages/components/src/button/stories/index.tsx:
```diff
@@ -13,9 +19,9 @@ import {
+const meta: ComponentMeta< typeof Button > = {
 	title: 'Components/Button',
 	component: Button,
 	argTypes: {
```
I think we can clean up a lot of the `argTypes` stuff in here now that the TS types will being generating them automatically.

kienstra replied:
Good idea! 

https://github.com/WordPress/gutenberg/pull/46997/commits/be1359d80500712d69ede04e10657a1b4beab81f removes all `argTypes` except `icon`, which has a useful `<select>`:

<img width="682" alt="Screen Shot 2023-01-21 at 10 33 59 PM" src="https://user-images.githubusercontent.com/4063887/213900791-1a42096d-e8ea-4840-a051-f3e1a41f4fc9.png">

mirka commented on packages/components/src/button/deprecated.tsx:
```diff
@@ -8,9 +12,23 @@ import { forwardRef } from '@wordpress/element';
+	}: WordPressComponentProps<
+		ButtonProps & DeprecatedIconButtonProps,
+		TagName
+	>,
```
How about we go a little more DRY here:

```suggestion
	}: React.ComponentPropsWithoutRef< typeof Button > &
		DeprecatedIconButtonProps,
```

kienstra replied:
Good idea! Committed in https://github.com/WordPress/gutenberg/pull/46997/commits/3632a42bc8eef7f6167a24bc2d2330881f35ad5c

mirka commented on packages/components/src/button/index.tsx:
```diff
@@ -17,8 +17,18 @@ import { useInstanceId } from '@wordpress/compose';
+const disabledEventsOnDisabledButton: DisabledEvent[] = [
+	'onMouseDown',
+	'onClick',
+];
```
```suggestion
const disabledEventsOnDisabledButton = [
	'onMouseDown',
	'onClick',
] as const;
```

We can just use `as const` here instead of declaring the type separately. It tells TS to lock the array from mutation, which then gives it permission to treat it like an enum.

kienstra replied:
That's a nice one! Committed in [309db5c](https://github.com/WordPress/gutenberg/pull/46997/commits/309db5cf29dc7dd2cc9291841223c53936a8443d#diff-d7a7249cdd571803b98fdd09650036566a355a402a68a58043a018acf0970ddbR28).

You showed me that before, sorry for missing it this time.

mirka reviewed (commented):
Major props for doing this one, @kienstra! It's one of the components I was absolutely dreading so I'm happy you went in and dealt with it head on 🔥⚔️

One big thing I'm wondering if we should do in this PR is to make the Button prop types a bit more strict, so a consumer would be prevented from mixing `button` and `a` props. Not that we have to use this diff verbatim, but [here's an example](https://github.com/WordPress/gutenberg/commit/165ef219c73378216e9197ce407442c4f655e6e1) to demonstrate how that could be done.

Here are some "test cases" that I have in mind:

```jsx
/**
 * Static TypeScript tests
 */
<Button href="foo" />;
// @ts-expect-error - `target` requires `href`
<Button target="foo" />;
// @ts-expect-error - `disabled` is only for buttons
<Button href="foo" disabled />;
<Button href="foo" type="image/png" />;
// @ts-expect-error - if button, type must be submit/reset/button
<Button type="image/png" />;
// @ts-expect-error
<Button type="invalidtype" />;
// @ts-expect-error - although the runtime behavior will allow this to be an anchor, this is probably a mistake so let's make it a TS error
<Button disabled __experimentalIsFocusable href="foo" />;
```

(For some complicated types like this one, I sometimes keep "static TS tests" like these in the unit test file so we can check that the types are working as expected.)

Let me know what you think — it's more like a nice-to-have, so it doesn't necessarily have to land in this PR.

kienstra committed:
Commit Lena's suggestion: Update packages/components/src/button/types.ts

Co-authored-by: Lena Morita <lena@jaguchi.com>

kienstra committed:
Commit Lena's suggestion: Update packages/components/src/button/deprecated.tsx

Co-authored-by: Lena Morita <lena@jaguchi.com>

kienstra committed:
Commit Lena's suggestion: Update packages/components/src/button/index.tsx

Co-authored-by: Lena Morita <lena@jaguchi.com>

kienstra committed:
Apply Lena's suggestion to change 'div' to 'button'

Also, correct some typing errors.

kienstra committed:
Remvoe needless type DisabledEvent

kienstra committed:
Remove needless argTypes that Storybook will infer from types

kienstra committed:
Restore a deleted test, thanks to Lena's idea

As Lena mentioned, the intent of this test looks like
ensuring that additional props are passed to the element.

kienstra committed:
Don't expect the console to error

kienstra committed:
Update tooltipPosition in README.md

kienstra committed:
Merge branch 'trunk' into update/button-ts

kienstra committed:
Move the CHANGELOG entry to Unreleased

kienstra committed:
Cherry-pick @mirka 's  commit to make Button types specific to a or button

kienstra committed:
Commit Lena's static tests verbatim

https://github.com/WordPress/gutenberg/pull/46997#pullrequestreview-1262338583

kienstra committed:
Fix a failed test run that I caused

kienstra committed:
Rename CommonButtonProps to BaseButtonProps

kienstra commented on packages/components/src/button/test/index.tsx:
```diff
@@ -360,33 +361,55 @@ describe( 'Button', () => {
+			// @ts-expect-error
+			render( <Button type="invalidtype" /> );
+			// @ts-expect-error - although the runtime behavior will allow this to be an anchor, this is probably a mistake.
+			render( <Button disabled __experimentalIsFocusable href="foo" /> );
```
If it's alright, this is basically copy-pasted from your https://github.com/WordPress/gutenberg/pull/46997#pullrequestreview-1262338583.

There's no `expect()` call, which is a little strange.

But it's not testing rendering, only typing.

mirka replied:
Right. Also it doesn't even need to be in a `render()` call 😄 I don't think there's an established convention for this kind of static TS test, so I'm just going freestyle (for example [these](https://github.com/search?q=repo%3AWordPress%2Fgutenberg%20%22static%20typescript%22&type=code)).

To take it further, it doesn't even need to be in `describe()` or `it()` calls, but I'm trying to stay somewhat in the realm of normal unit tests to at least signal that it's some kind of test, just so other devs aren't overly confused.

kienstra replied:
Yeah, it'd definitely be simpler to not wrap them in `render()` and `it()`.

I can't get that to work, without getting 'JSX expressions must have one parent element':
<img width="887" alt="Screen Shot 2023-01-23 at 5 55 41 PM" src="https://user-images.githubusercontent.com/4063887/214179246-38ae822a-ce87-4279-8edd-26efade36922.png">

```jsx
	describe( 'static typing', () => {
		<Button href="foo" />

		// @ts-expect-error - `target` requires `href`
		<Button target="foo" />

		// @ts-expect-error - `disabled` is only for buttons
		<Button href="foo" disabled />

		<Button href="foo" type="image/png" />

		// @ts-expect-error - if button, type must be submit/reset/button
		<Button type="image/png" />;

		// @ts-expect-error
		<Button type="invalidtype" />

		// @ts-expect-error - although the runtime behavior will allow this to be an anchor, this is probably a mistake.
		<Button disabled __experimentalIsFocusable href="foo" />
	} );
```

The examples you [linked to](https://github.com/search?q=repo%3AWordPress%2Fgutenberg%20%22static%20typescript%22&type=code) work, though 😄

mirka replied:
Lol. We can use a Fragment, I guess?

```jsx
describe( 'static typing', () => {
	<>
		<Button href="foo" />
		{ /* @ts-expect-error - `target` requires `href` */ }
		<Button target="foo" />
		{ /* @ts-expect-error - `disabled` is only for buttons */ }
		<Button href="foo" disabled />
		<Button href="foo" type="image/png" />
		{ /* @ts-expect-error - if button, type must be submit/reset/button */ }
		<Button type="image/png" />
		{ /* @ts-expect-error */ }
		<Button type="invalidtype" />
		{ /* @ts-expect-error - although the runtime behavior will allow this to be an anchor, this is probably a mistake. */ }
		<Button disabled __experimentalIsFocusable href="foo" />
	</>;
} );
```

kienstra replied:
Ah, good idea.

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/46997/commits/dfef117f1707f9a24c519f04832799633991aa54

kienstra commented:
>One big thing I'm wondering if we should do in this PR is to make the Button prop types a bit more strict, so a consumer would be prevented from mixing button and a props. Not that we have to use this diff verbatim, but [here's an example](https://github.com/WordPress/gutenberg/commit/165ef219c73378216e9197ce407442c4f655e6e1) to demonstrate how that could be done.

Good idea! I cherry-picked your https://github.com/WordPress/gutenberg/commit/165ef219c73378216e9197ce407442c4f655e6e1 with only a [name change](https://github.com/WordPress/gutenberg/pull/46997/commits/3892543641724b190a86ddc6064af59ceebd0483), and copy-pasted your [static TypeScript tests](https://github.com/WordPress/gutenberg/commit/ccd87bcdb69ff1ea60e660bfe472dd4e31401f96).

kienstra commented:
Hi @mirka,
Just finished applying your suggestions.

Thanks for all the care and code you put into this.

kienstra committed:
Merge in trunk, resolve conflict

mirka commented on packages/components/src/button/stories/index.tsx:
```diff
@@ -0,0 +1,104 @@
+const meta: ComponentMeta< typeof Button > = {
+	title: 'Components/Button',
+	component: Button,
+	argTypes: {
```
One last thing for the `argTypes`, let's make this small override so the `href` prop doesn't look like a required prop in the Storybook props table:

```suggestion
	argTypes: {
		// Overrides a limitation of the docgen interpreting our TS types for this as required.
		href: { type: { name: 'string', required: false } },
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/46997/commits/07b70022b95ef031a4e3194f49fda1ba64ef4266

mirka reviewed (approved):
Really great work here. I'm happy we finally have types for possibly the single most widely used component in the package!

kienstra committed:
Commit Lena's suggestion: Update packages/components/src/button/stories/index.tsx

Co-authored-by: Lena Morita <lena@jaguchi.com>

kienstra commented:
>Really great work here. I'm happy we finally have types for possibly the single most widely used component in the package!

Thanks so much, Lena! Just responded to both of your comments.

kienstra committed:
Commit Lena's diff for static typing test

kienstra commented:
Going to merge this in 30 minutes.

ciampo commented:
Just passing by to thank you, Ryan! We really appreciate your contributions to the project and the components library 🚀

kienstra commented:
Thanks so much, Marco!
